README.markdown
twmn
====
A notification system for tiling window managers. `twmn` is two things:

`twmnc`: command line tool to send notifications to `twmnd`. You can also use `notify-send` for a similar purpose, but `twmnc` is more powerful. See `twmnc --help` for more information.

`twmnd`: daemon listening to notification requests and showing them one after another. Configure it at `~/.twmn`. The file is generated the first time `twmnd` is launched.

Notifications are shown in a one-line bar called the notification slide. They can be navigated through and activated with shortcuts.

See `.twmn` for more information.


About .twmn
---------------
```ini
[gui]
; Screen number to display notifications on when using a multi-head desktop.
screen=  ; 0 indexed screen number

; WARNING: Deprecated by "screen"
; Absolute position from the top-left corner of the slide. You may need it for a multi-screen setup.
; You still have to set position in order to choose the slide animation. If empty, twmnd will try
; to figure out where to display the slide according to your desktop size and the slide position.
absolute_position=  ; Supported format: WxH. Width and Height being integers.

; Background color.
background_color=black  ; RBG hex and keywords (eg. lightgray) are supported.

; An icon for the layout. Useful only for a layout file.
icon=  ; Path to image file. Optional.

; Font family.
font=Sans

; Font size.
font_size=13  ; In pixel.

; Text color.
foreground_color=white  ; RBG hex and keywords (eg. lightgray) are supported.

; Height of the slide bar. Useful to match your tiling window manager's bar.
height=18  ; In pixel.

; Position of the notification slide.
position=top_right  ; Accepted values: top_right (tr), top_left (tl), bottom_right (br),
                    ; bottom_left (bl), top_center (tc), bottom_center (bc), center (c).

; The animation to use when the slide appear
in_animation=38 ; see http://doc.qt.nokia.com/latest/qeasingcurve.html#Type-enum for types

; The in animation's duration
in_animation_duration=1000 ; in milliseconds

; The animation to use whe the slide is closing
out_animation=13

; The out animation's duration
out_animation_duration=1000

; Enable or disable notification bounce when changing notification
bounce=true  ; true or false

[icons]
; An icon. You can add as many as you want.
NAME=  ; Path to image file. NAME being the icon's custom name.


[main]
; Program/command to be executed on notification activation.
activate_command=  ; Path to command.

; Time each notification remains visible.
duration=3000  ; In millisecond.

; Enable or disable shortcuts.
enable_shortcuts=true  ; true or false.

; UDP port used for notifications.
port=9797

; Program/command to play sound with.
sound_command=  ;  Path to command. Leave empty for no sound.


[shortcuts]
; Modifiers shortcuts.
modifiers=Alt+  ; Up to three modifiers. Use with the following shortcuts.

; Shows the previous notification. Mouse wheel up does it too.
previous=K

; Shows the next notification. If a modification is manually shown it will not be displayed again
; when twmnd process the notification stack. Mouse wheel up does it too.
next=J

; Activates the notification. Runs the command defined at activate_command.
; Mouse left click activates it too.
activate=Return

; Hides all notifications. To hide the current notification, use the "next" key
; instead. Mouse right click has the same effect.
hide=X
```


Installation
------------

For [Arch Linux](http://www.archlinux.org/) users, `twmn` is [on the AUR](https://aur.archlinux.org/packages.php?ID=51596).

For [Debian](https://www.debian.org/) or [Ubuntu](http://www.ubuntu.com/) you can (for now) build the package using your preferred build-tool (eg. `dpkg-buildpackage`) for building as a debian package. Install using `dpkg -i`

Otherwise you can install `twmnd` and `twmnc` manually:

1. install `boost`, `qt` and `dbus` if they weren't before
2. `git clone https://github.com/sboli/twmn.git` to get `twmn`
3. `cd twmn/`
4. `qmake` to generate a Makefile
5. `make` to compile
6. `sudo make install` to install `twmnd` and `twmnc` in `/usr/local/bin`. Make sure this folder is in your `$PATH` environment variable. (`export PATH=$PATH:/usr/local/bin`)
7. for `twmnd` to be always running, add it to your `.xinitrc`, `rc.conf` or else

The `storage_notifier` example requires `dbus-python` to be installed. The `mpd_notifier` example requires `python-mpd` to be installed and running.


debian/changelog
twmn (0.1.0) UNRELEASED; urgency=low

  * Initial release.

 -- Morten Hekkvang <morten.hekkvang@gmail.com>  Fri, 21 Feb 2014 13:53:49 +0100


twmnc/main.cpp
#include <iostream>
#include <fstream>
#include <boost/program_options.hpp>
#include <boost/asio.hpp>
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/xml_parser.hpp>
#include <boost/property_tree/ini_parser.hpp>
#include <boost/tuple/tuple.hpp>
#include <boost/lambda/lambda.hpp>
#include <boost/foreach.hpp>
#include <exception>
#include <unistd.h>

const std::string   DEFAULT_HOST = "127.0.0.1";
const int           DEFAULT_PORT = 9797;

/*!
  * \brief Read the port from the config file
  * \return true if the port was found
  * \param port the port to use
  */
bool read_port(int& port)
{
    std::string path = getenv("HOME") ;
    path += "/.twmn";
    std::ifstream in(path.c_str());
    if (!in)
        return false;
    boost::property_tree::ptree tree;
    boost::property_tree::ini_parser::read_ini(in, tree);
    boost::optional<int> value =  tree.get_optional<int>("main.port");
    return value ? port = *value : false;
}

/*!
  * \brief Fill tree with command line options from vm
  * \return true if any command line options was found
  */
bool populate_tree(boost::program_options::variables_map& vm, boost::property_tree::ptree& tree)
{
    boost::program_options::notify(vm);
    boost::property_tree::ptree& root = tree.add("root", "");
    boost::property_tree::ptree empty_root = root;
    BOOST_FOREACH(boost::program_options::variables_map::value_type& i, vm) {
        boost::any value = i.second.value();
        try {
            root.add(i.first, boost::any_cast<std::string>(value));
        }
        catch (const boost::bad_any_cast& e) {
            root.add(i.first, boost::any_cast<int>(value));
        }
    }
    return empty_root != root; // tree not empty
}

/*!
  * \brief Sent tree to the host
  */
void send_tree(const boost::program_options::variables_map& vm, const boost::property_tree::ptree& tree)
{
    using namespace boost::asio;
    std::ostringstream oss;
    boost::property_tree::xml_parser::write_xml(oss, tree);
    io_service ios;
    ip::udp::socket s(ios, ip::udp::endpoint(ip::udp::v4(), 0));
    int port = vm.count("port") ? vm["port"].as<int>() : 0;
    if (!port)
        if (!read_port(port))
            port = DEFAULT_PORT;
    boost::optional<std::string> host = tree.get_optional<std::string>("content.host");
    if (!host)
        host = boost::optional<std::string>(DEFAULT_HOST);
    s.send_to(buffer(oss.str()), ip::udp::endpoint(ip::address(ip::address_v4::from_string(*host)), port));
    ios.run();
}

int main(int argc, char** argv)
{
    namespace po = boost::program_options;
    po::options_description desc("Options");
    desc.add_options()
            ("help,h", "Show this help")
            ("icon,i", po::value<std::string>(), "The icon to use, either a predefined symbol"
                                                  " in the config file or a file path.")
            ("title,t", po::value<std::string>(), "A title for the notification.")
            ("content,c", po::value<std::string>(), "A content for this notification")
            ("host,H", po::value<std::string>(), ("The host ip address (x.x.x.x). Default is " + DEFAULT_HOST).c_str())
            ("port,p", po::value<int>(), "The port to use. Default setting is loaded from the config file.")
            ("layout,l", po::value<std::string>(), "The layout to use. The name of a configuration file.")
            ("size,s", po::value<int>(), "The height of this notification. If not specified the configuration file value is used.")
            ("pos", po::value<std::string>(), "Position of this notification. Either top_right, top_left, bottom_right or bottom_left.")
            ("fn", po::value<std::string>(), "The font to use for this notification.")
            ("fs", po::value<std::string>(), "The font size for this notification.")
            ("duration,d", po::value<int>(), "How long this notification remains visible")
            ("sc", po::value<std::string>(), "A command to play a specific sound for this notification")
            ("bg", po::value<std::string>(), "The background color")
            ("fg", po::value<std::string>(), "The foreground color, which is the font color.")
            ("id", po::value<int>(), "A message id. You should have given an id to the first message in order to modify it")
            ("aot", "Always on top. Specially on fullscreen applications, default.")
            ("ac", po::value<std::string>(), "A command to run when the notification is activated.")
    ;
    try {
        po::variables_map vm;
        boost::property_tree::ptree tree;
        po::store(po::parse_command_line(argc, argv, desc), vm);
        if (vm.count("help")) {
	    std::cout << desc << std::endl;
	    return 0;
	}
        const bool filled = populate_tree(vm, tree);
        if (!filled && argc > 1) {
            std::string full_line;
            std::for_each(argv+1, argv+argc, (full_line += boost::lambda::_1) += " ");
            full_line.erase(full_line.length()-1);
            tree.put("root.content", full_line);
        }
        else if (!filled && argc == 1)
            throw std::runtime_error("Empty command line");
        send_tree(vm, tree);
    }
    catch (const std::exception& e) {
        std::cout << desc << std::endl;
    }
    return 0;
}


twmnd/settings.h
#ifndef SETTINGS_H
#define SETTINGS_H

#include <QSettings>
#include <QApplication>
#include <QtCore>

/*!
  * \brief Gère les paramètres du programme
  */
class Settings
{
public:
    Settings(QString file = QString(""));

    /*!
      * \brief Sauvegarde les réglages
      */
    ~Settings();

    /*!
      * \brief Set default values.
      */
    void                setDefaults();

    /*!
      * \brief Recharge les paramètres à partir du fichier.
      */
    void                reload();

    /*!
      * \brief Enregistre le paramètre setting
      */
    void                set(QString setting, const QVariant& value);

    /*!
      * \brief Récupère le paramètre setting
      */
    QVariant            get(QString setting);

    /*!
      * \return True si le paramètre setting existe.
      */
    bool                has(QString setting);

    /*!
      * \brief Sauvegarde les paramètres.
      */
    void                save();

    /*!
      * \brief Complète cette instance avec les donnés de s, sans écraser les données de cette instance.
      */
    void                fillWith(const Settings& s);

private:
    QMap<QString, QVariant> m_data;
    QMap<QString, QVariant> m_defaultSettings;
    QString                 m_file;
};

#endif // SETTINGS_H


twmnd/settings.cpp
#include "settings.h"
#include <iostream>
#include <QKeySequence>

Settings::Settings(QString file)
{
    m_file = QString(std::getenv("HOME")) + QString("/.twmn");

    if (!file.isEmpty()) {
        m_file += QString("_") + file;
    }

    setDefaults();
    reload();
}

void Settings::setDefaults()
{
    m_defaultSettings["main/port"] = 9797;
    m_defaultSettings["main/sound_command"] = "";
    m_defaultSettings["main/activate_command"] = "";
    m_defaultSettings["main/duration"] = "3000";
    m_defaultSettings["main/enable_shortcuts"] = true;
    m_defaultSettings["gui/position"] = "top_right";
    m_defaultSettings["gui/screen"] = "";
    m_defaultSettings["gui/absolute_position"] = "";
    m_defaultSettings["gui/bounce"] = true;
    m_defaultSettings["gui/height"] = 18;
    m_defaultSettings["gui/font"] = "Sans";
    m_defaultSettings["gui/font_size"] = 13;
    m_defaultSettings["gui/foreground_color"] = "#999999";
    m_defaultSettings["gui/background_color"] = "#000000";
    m_defaultSettings["gui/always_on_top"] = true;
    m_defaultSettings["gui/in_animation"] = 38;
    m_defaultSettings["gui/in_animation_duration"] = 1000;
    m_defaultSettings["gui/out_animation"] = 13;
    m_defaultSettings["gui/out_animation_duration"] = 1000;
    m_defaultSettings["icons/critical_icon"] = "";
    m_defaultSettings["icons/warning_icon"] = "";
    m_defaultSettings["icons/info_icon"] = "";
    m_defaultSettings["shortcuts/modifiers"] = QKeySequence(Qt::AltModifier);
    m_defaultSettings["shortcuts/previous"] = QKeySequence("K");
    m_defaultSettings["shortcuts/next"] = QKeySequence("J");
    m_defaultSettings["shortcuts/activate"] = QKeySequence("Return");
    m_defaultSettings["shortcuts/hide"] = QKeySequence("X");
}

Settings::~Settings()
{
    save();
}

void Settings::reload()
{
    QSettings settings(m_file, QSettings::IniFormat);
    settings.setIniCodec("UTF-8");
    QStringList keys = settings.allKeys();
    m_data.clear();
    foreach (QString i, keys) {
        m_data[i] = settings.value(i);
    }
    save();
}

void Settings::set(QString setting, const QVariant &value)
{
    m_data[setting] = value;
}

QVariant Settings::get(QString setting)
{
    if (!m_data.contains(setting)) {
        if (!m_defaultSettings.contains(setting)) {
            std::cout << ("Attempt to get a non existing setting.\n"
                          "You might want to create a key for " + setting + " in the config file.").toStdString();
            qApp->quit();
        }
        return m_defaultSettings[setting];
    }
    return m_data[setting];
}

bool Settings::has(QString setting)
{
    return m_data.contains(setting) || m_defaultSettings.contains(setting);
}

void Settings::save()
{
    QSettings settings(m_file, QSettings::IniFormat);
    settings.setIniCodec("UTF-8");
    settings.clear();
    for (QMap<QString, QVariant>::const_iterator it = m_data.begin(); it != m_data.end(); ++it) {
        m_defaultSettings[it.key()] = it.value();
    }
    for (QMap<QString, QVariant>::const_iterator it = m_defaultSettings.begin(); it != m_defaultSettings.end(); ++it) {
        settings.setValue(it.key(), it.value());
    }
}

void Settings::fillWith(const Settings &s)
{
    for (QMap<QString, QVariant>::const_iterator it = s.m_data.begin(); it != s.m_data.end(); ++it) {
        if (m_data.contains(it.key()))
            continue;
        else
            m_data[it.key()] = it.value();
    }
}
